File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/regular_expression_matching/RegularExpressionMatching.java
package main.java.com.jiang.leetcode100.regular_expression_matching;

import java.util.Scanner;

class Solution {
    public boolean isMatch(String s, String p) {
        System.out.println(s);
        System.out.println(p);

        if(p.isEmpty()) {
            if(s.isEmpty()) return true;
            else return false;
        }

        if(p.length() == 1) {
            if(s.length() != 1) return false;
            if(p.charAt(0) == '.' || p.charAt(0) == s.charAt(0)) return true;
            return false;
        }

        if(p.charAt(1) != '*') {
            if(s.isEmpty()) return false;
            if(s.charAt(0) == p.charAt(0) || p.charAt(0) == '.') {
                return isMatch(s.substring(1), p.substring(1));
            }
            else return false;
        }
        
        for(int i = -1; i < s.length(); i++) {
            if(i == -1 || p.charAt(0) == '.' || s.charAt(i) == p.charAt(0)) {
                if(isMatch(s.substring(i + 1), p.substring(2))) {
                    return true;
                }
            } else break;
        }
        
        return false;
    }
}

public class RegularExpressionMatching {
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String s = scanner.nextLine();
        String p = scanner.nextLine();
        scanner.close();
        if(new Solution().isMatch(s, p)) {
            System.out.println("True");
        } else {
            System.out.println("False");
        }
    }
}

File: Leetcode100/leetcode100_python/RegularExpressionMatching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        if not p:
            return not s
        if len(p) == 1:
            if len(s) == 1 and (s[0] == p[0] or p[0] == '.'):
                return True
            return False
        if p[1] == '*':
            if len(s) > 0 and (s[0] == p[0] or p[0] == '.'):
                return self.isMatch(s, p[2: ]) or self.isMatch(s[1: ], p)
            else:
                return self.isMatch(s, p[2: ])
        if len(s) > 0 and (s[0] == p[0] or p[0] == '.'):
            return self.isMatch(s[1: ], p[1:])
        else:
            return False

if __name__ == "__main__":
    s = input()
    p = input()
    solution = Solution()
    print(solution.isMatch(s, p))